bindings/rabbitmq/rabbitmq.go

```diff
@@ -84,34 +96,78 @@ func (r *RabbitMQ) Init(_ context.Context, metadata bindings.Metadata) error {
 		return err
 	}
 
-	conn, err := amqp.Dial(r.metadata.Host)
+	err = r.connect()
 	if err != nil {
 		return err
 	}
+	r.notifyRabbitChannelClose = make(chan *amqp.Error, 1)
+	r.channel.NotifyClose(r.notifyRabbitChannelClose)
+	go r.reconnectWhenNecessary()
+	return nil
+}
 
-	ch, err := conn.Channel()
-	if err != nil {
-		return err
+func (r *RabbitMQ) reconnectWhenNecessary() {
+	for {
+		select {
+		case <-r.closeCh:
+			return
+		case _ = <-r.notifyRabbitChannelClose:
+			// this is called when the server restart or the channel is closed by server.
+			if r.connection != nil && !r.connection.IsClosed() {
+				ch, err := r.connection.Channel()
+				if err == nil {
+					r.channel = ch
+					r.notifyRabbitChannelClose = make(chan *amqp.Error, 1)
+					r.channel.NotifyClose(r.notifyRabbitChannelClose)
+					continue
+				}
+				// if encounter err fallback to reconnect connection
+			}
+			// keep trying to reconnect
+			for {
+				err := r.connect()
+				if err == nil {
+					break
+				}
+				if err == errClosed {
+					return
+				}
+				r.logger.Warnf("reconnect failed: %s", err.Error())
+
+				time.Sleep(r.metadata.ReconnectWait)
+			}
+
+			r.notifyRabbitChannelClose = make(chan *amqp.Error, 1)
+			r.channel.NotifyClose(r.notifyRabbitChannelClose)
+		}
 	}
-	ch.Qos(r.metadata.PrefetchCount, 0, true)
-	r.connection = conn
-	r.channel = ch
+}
+
+func dial(uri string) (conn *amqp.Connection, ch *amqp.Channel, err error) {
+	conn, err = amqp.Dial(uri)
 
-	q, err := r.declareQueue()
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 
-	r.queue = q
+	ch, err = conn.Channel()
+	if err != nil {
+		conn.Close()
+		return nil, nil, err
+	}
 
-	return nil
+	return conn, ch, nil
 }
 
 func (r *RabbitMQ) Operations() []bindings.OperationKind {
 	return []bindings.OperationKind{bindings.CreateOperation}
 }
 
 func (r *RabbitMQ) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// check if connection channel to rabbitmq is open
+	if r.channel == nil {
```

Comment:
Ditto. This probably needs to be wrapped in a lock or it can cause race conditions

Reply:
Done.